src/overlords/offense/roomPoisoner.ts

```diff
@@ -1,96 +1,99 @@
-//
-// 				const dismantleTarget = this.findStructureBlockingController(roomPoisoner);
-// 				if (dismantleTarget) {
-// 					roomPoisoner.task = Tasks.dismantle(dismantleTarget);
-// 					return;
-// 				}
-// 			}
-//
-//
-// 			// recharge
-// 			if (roomPoisoner.carry.energy == 0) {
-// 				roomPoisoner.task = Tasks.recharge();
-// 			} else if (this.room && this.room.controller &&
-// 					   (this.room.controller.level < 2) &&
-// 					   !(this.room.controller.upgradeBlocked > 0)) {
-// 				// upgrade controller to level 2 to unlock walls
-// 				roomPoisoner.task = Tasks.upgrade(this.room.controller);
-// 			} else if (this.controllerWallSites && this.controllerWallSites.length) {
-// 				roomPoisoner.task = Tasks.build(this.controllerWallSites[0]);
-// 			} else if (this.sourcesWallSites && this.sourcesWallSites.length) {
-// 				roomPoisoner.task = Tasks.build(this.sourcesWallSites[0]);
-// 			}
-// 		} else {
-// 			roomPoisoner.goTo(this.pos, {ensurePath: true, avoidSK: true});
-// 		}
-// 	}
-//
-// 	run() {
-// 		this.autoRun(this.roomPoisoners, roomPoisoner => this.handleRoomPoisoner(roomPoisoner));
-// 	}
-// }
-//
+import {Roles, Setups} from '../../creepSetups/setups';
+import {DirectivePoisonRoom} from '../../directives/offense/poisonRoom';
+import {OverlordPriority} from '../../priorities/priorities_overlords';
+import {profile} from '../../profiler/decorator';
+import {Tasks} from '../../tasks/Tasks';
+import {minBy} from '../../utilities/utils';
+import {Zerg} from '../../zerg/Zerg';
+import {Overlord} from '../Overlord';
+
+export const MINIMUM_WALL_HITS = 1;
+
+/**
+ * Spawn roomPoisoner - upgrqde controller to lvl2, wall in controller then sources.
+ */
+@profile
+export class RoomPoisonerOverlord extends Overlord {
+
+	directive: DirectivePoisonRoom;
+	roomPoisoners: Zerg[];
+
+	constructor(directive: DirectivePoisonRoom, priority = OverlordPriority.offense.roomPoisoner) {
+		super(directive, 'PoisonRoom', priority);
+
+		this.directive = directive;
+		this.roomPoisoners = this.zerg(Roles.roomPoisoner);
+	}
+
+	init() {
+		if(this.room && this.room.dangerousPlayerHostiles.length == 0) {
+			this.wishlist(1, Setups.roomPoisoner);
+		}
+	}
+	
+	private handleRoomPoisoner(roomPoisoner: Zerg): void {
+		// Recharge from colony room.
+		if(roomPoisoner.inSameRoomAs(this.colony) && roomPoisoner.carry.energy == 0) {
+			roomPoisoner.task = Tasks.recharge();
+			return;
+		}
+		// Go to Target Room
+		if (!roomPoisoner.inSameRoomAs(this.directive)) {
+			roomPoisoner.goTo(this.pos, {ensurePath: true, avoidSK: true});
+			return;
+		}
+		// all actions below are done in target directive room
+		// recharge in target room
+		if (roomPoisoner.carry.energy == 0) {
+			roomPoisoner.task = Tasks.recharge();
+			return;
+		}
+
+		// upgrade controller to level 2
+		if(this.room && this.room.controller && this.room.controller.level < 2) {
+			roomPoisoner.task = Tasks.upgrade(this.room.controller);
+			return;
+		}
+
```

Comment:
use if, else if, etc for these instead of if's and returning. Remove the tons of duplicate room and controller checks

if (roomPoisoner.carry.energy == 0) {
			roomPoisoner.task = Tasks.recharge();
			return;
		} else if (controller.level <2) {

Reply:
well, this makes it much easier for me to read!
I wrapped them around if(this.room && this.room.controller) {
check it out and let me know